labs/CryptoProofs/CryptoProofs.md

```diff
@@ -0,0 +1,232 @@
+(\x -> square x == 1764) 42 = True
+(Total Elapsed Time: 0.021s, using "Z3")
+```
+
+Let's take a closer look at this query, which makes use of a *lambda* (anonymous/unnamed/on-the-fly) function. Here's the breakdown:
+
+|Function Reversal||||
+|-|-|-|-|
+| `:sat`            | `\x`                   | `->`        | `square x == 1764` |
+| "Hey SAT solver!" | "Find me an input `x`" | "such that" | "`x` squared equals `1764`" |
+|||||
+
+### Exercise 2.1.1 Reverse DES.encrypt
+
+Given the following key and ciphertext, find the plaintext using only the solver and the `DES.encrypt` function.
+
+```
+known_key = 0x752979387592cb70
+known_ct = 0xf2930290ea4db580
+```
+
+Note: For whatever reason, the default Z3 solver has trouble with this one. Try one of the other solvers, such as yices:
+
+```bash
+labs::CryptoProofs::CryptoProofs> :s prover=yices
+```
+
+Or use all the installed solvers in a first-to-the-post race.
+*Caution! May exhaust system resources.*
+
+```bash
+labs::CryptoProofs::CryptoProofs> :s prover=any
+```
+
+### Exercise 2.1.2 Breaking DES
+
+Given the following matched plaintext and ciphertext, ask the solver to find the key. Will this work? Why or why not? (*Hint: see plaintext.*) Note that you can stop the solver at any time by hitting `ctrl-c`.
+
+```
+matched_pt = join "tootough"
+matched_ct = 0x95d07f8a72707733
+```
+
+To make this solvable, try it again with the first six bytes of key provided: `0x1234567890ab`.
+
+## 2.2 Proof of Inversion
+
+For symmetric ciphers, it is necessary that the decrypt function *inverts* the encrypt function. (That is, it restores the ciphertext to the original plaintext.) It is easy to express this property in Cryptol.
+
+Consider the following functions `f` and `g`:
+
+```
+f: Integer -> Integer
+f x = 3 * x + 2
+g: Integer -> Integer
+g x = (x - 2) / 3
+```
+
+We want to prove that function `g` inverts function `f`; that is, applying `g` to the result of `f x` gets `x` back. Here's the invocation:
+
+```bash
+labs::CryptoProofs::CryptoProofs> :prove \x -> g (f x) == x
+Q.E.D.
+(Total Elapsed Time: 0.023s, using "Z3")
+```
+
+Here's the breadown of this proof:
+
+|Proof of Inversion||||
+|-|-|-|-|
+| `:prove`            | `\x`                   | `->`        | `g (f x) == x` |
+| "Prove to me" | "that for all `x`" | "it is true that" | "`g` inverts `f`" |
+|
+
+### Exercise 2.2.1 The other direction
+
+Our example proof showed that `g` inverts `f` for all inputs. Does this work the other way around? Try it! If the proof fails, it will provide a *counterexample*. Use the counterexample to understand what went wrong.
+
+### Exercise 2.2.2 DES inversion
+
+Use Cryptol to prove that `DES.encrypt` and `DES.decrypt` are inverses for all possible inputs. Show both directions.
+
+*Hint*: Note that this function has more than one input. To make it work, you can either use a tuple `(,)` (tuple-style) or a chain of of `\->` (Curried-style) constructs for your lambda function input.
```

Comment:
Did you know that this style also works? I had no idea until today!

```
:prove \x y -> x + y == (0 : [32])
```

Reply:
Didn't know! I'll update.